// drizzle-orm/session-805d2876.cjs
'use strict';

var alias = require('./alias-3f5ac719.cjs');
require('./index.cjs');

var _a$d, _b$9, _c$5;
/** @internal */
const InlineForeignKeys = Symbol.for('drizzle:MySqlInlineForeignKeys');
class MySqlTable extends alias.Table {
    constructor() {
        super(...arguments);
        /** @internal */
        this[_b$9] = [];
        /** @internal */
        this[_c$5] = undefined;
    }
}
_a$d = alias.entityKind, alias.Table.Symbol.Columns, _b$9 = InlineForeignKeys, _c$5 = alias.Table.Symbol.ExtraConfigBuilder;
MySqlTable[_a$d] = 'MySqlTable';
/** @internal */
MySqlTable.Symbol = Object.assign({}, alias.Table.Symbol, {
    InlineForeignKeys: InlineForeignKeys,
});
function mysqlTableWithSchema(name, columns, extraConfig, schema, baseName = name) {
    const rawTable = new MySqlTable(name, schema, baseName);
    const builtColumns = Object.fromEntries(Object.entries(columns).map(([name, colBuilder]) => {
        const column = colBuilder.build(rawTable);
        rawTable[InlineForeignKeys].push(...colBuilder.buildForeignKeys(column, rawTable));
        return [name, column];
    }));
    const table = Object.assign(rawTable, builtColumns);
    table[alias.Table.Symbol.Columns] = builtColumns;
    if (extraConfig) {
        table[MySqlTable.Symbol.ExtraConfigBuilder] = extraConfig;
    }
    return table;
}
const mysqlTable = (name, columns, extraConfig) => {
    return mysqlTableWithSchema(name, columns, extraConfig, undefined, name);
};
function mysqlTableCreator(customizeTableName) {
    return (name, columns, extraConfig) => {
        return mysqlTableWithSchema(customizeTableName(name), columns, extraConfig, undefined, name);
    };
}

var _a$c, _b$8;
class ForeignKeyBuilder {
    constructor(config, actions) {
        this.reference = () => {
            const { columns, foreignColumns } = config();
            return { columns, foreignTable: foreignColumns[0].table, foreignColumns };
        };
        if (actions) {
            this._onUpdate = actions.onUpdate;
            this._onDelete = actions.onDelete;
        }
    }
    onUpdate(action) {
        this._onUpdate = action;
        return this;
    }
    onDelete(action) {
        this._onDelete = action;
        return this;
    }
    /** @internal */
    build(table) {
        return new ForeignKey(table, this);
    }
}
_a$c = alias.entityKind;
ForeignKeyBuilder[_a$c] = 'MySqlForeignKeyBuilder';
class ForeignKey {
    constructor(table, builder) {
        this.table = table;
        this.reference = builder.reference;
        this.onUpdate = builder._onUpdate;
        this.onDelete = builder._onDelete;
    }
    getName() {
        const { columns, foreignColumns } = this.reference();
        const columnNames = columns.map((column) => column.name);
        const foreignColumnNames = foreignColumns.map((column) => column.name);
        const chunks = [
            this.table[MySqlTable.Symbol.Name],
            ...columnNames,
            foreignColumns[0].table[MySqlTable.Symbol.Name],
            ...foreignColumnNames,
        ];
        return `${chunks.join('_')}_fk`;
    }
}
_b$8 = alias.entityKind;
ForeignKey[_b$8] = 'MySqlForeignKey';
function foreignKey(config) {
    function mappedConfig() {
        const { columns, foreignColumns } = config;
        return {
            columns,
            foreignColumns,
        };
    }
    return new ForeignKeyBuilder(mappedConfig);
}

var _a$b, _b$7, _c$4;
function unique(name) {
    return new UniqueOnConstraintBuilder(name);
}
function uniqueKeyName(table, columns) {
    return `${table[MySqlTable.Symbol.Name]}_${columns.join('_')}_unique`;
}
class UniqueConstraintBuilder {
    constructor(columns, name) {
        this.name = name;
        this.columns = columns;
    }
    /** @internal */
    build(table) {
        return new UniqueConstraint(table, this.columns, this.name);
    }
}
_a$b = alias.entityKind;
UniqueConstraintBuilder[_a$b] = 'MySqlUniqueConstraintBuilder';
class UniqueOnConstraintBuilder {
    constructor(name) {
        this.name = name;
    }
    on(...columns) {
        return new UniqueConstraintBuilder(columns, this.name);
    }
}
_b$7 = alias.entityKind;
UniqueOnConstraintBuilder[_b$7] = 'MySqlUniqueOnConstraintBuilder';
class UniqueConstraint {
    constructor(table, columns, name) {
        this.table = table;
        this.nullsNotDistinct = false;
        this.columns = columns;
        this.name = name ?? uniqueKeyName(this.table, this.columns.map((column) => column.name));
    }
    getName() {
        return this.name;
    }
}
_c$4 = alias.entityKind;
UniqueConstraint[_c$4] = 'MySqlUniqueConstraint';

var _a$a, _b$6, _c$3, _d$1;
class MySqlColumnBuilder extends alias.ColumnBuilder {
    constructor() {
        super(...arguments);
        this.foreignKeyConfigs = [];
    }
    references(ref, actions = {}) {
        this.foreignKeyConfigs.push({ ref, actions });
        return this;
    }
    unique(name) {
        this.config.isUnique = true;
        this.config.uniqueName = name;
        return this;
    }
    /** @internal */
    buildForeignKeys(column, table) {
        return this.foreignKeyConfigs.map(({ ref, actions }) => {
            return ((ref, actions) => {
                const builder = new ForeignKeyBuilder(() => {
                    const foreignColumn = ref();
                    return { columns: [column], foreignColumns: [foreignColumn] };
                });
                if (actions.onUpdate) {
                    builder.onUpdate(actions.onUpdate);
                }
                if (actions.onDelete) {
                    builder.onDelete(actions.onDelete);
                }
                return builder.build(table);
            })(ref, actions);
        });
    }
}
_a$a = alias.entityKind;
MySqlColumnBuilder[_a$a] = 'MySqlColumnBuilder';
// To understand how to use `MySqlColumn` and `AnyMySqlColumn`, see `Column` and `AnyColumn` documentation.
class MySqlColumn extends alias.Column {
    constructor(table, config) {
        if (!config.uniqueName) {
            config.uniqueName = uniqueKeyName(table, [config.name]);
        }
        super(table, config);
        this.table = table;
    }
}
_b$6 = alias.entityKind;
MySqlColumn[_b$6] = 'MySqlColumn';
class MySqlColumnBuilderWithAutoIncrement extends MySqlColumnBuilder {
    constructor(name, dataType, columnType) {
        super(name, dataType, columnType);
        this.config.autoIncrement = false;
    }
    autoincrement() {
        this.config.autoIncrement = true;
        this.config.hasDefault = true;
        return this;
    }
}
_c$3 = alias.entityKind;
MySqlColumnBuilderWithAutoIncrement[_c$3] = 'MySqlColumnBuilderWithAutoIncrement';
class MySqlColumnWithAutoIncrement extends MySqlColumn {
    constructor() {
        super(...arguments);
        this.autoIncrement = this.config.autoIncrement;
    }
}
_d$1 = alias.entityKind;
MySqlColumnWithAutoIncrement[_d$1] = 'MySqlColumnWithAutoIncrement';

var _a$9;
class MySqlDelete extends alias.QueryPromise {
    constructor(table, session, dialect) {
        super();
        this.table = table;
        this.session = session;
        this.dialect = dialect;
        this.execute = (placeholderValues) => {
            return this.prepare().execute(placeholderValues);
        };
        this.createIterator = () => {
            const self = this;
            return async function* (placeholderValues) {
                yield* self.prepare().iterator(placeholderValues);
            };
        };
        this.iterator = this.createIterator();
        this.config = { table };
    }
    where(where) {
        this.config.where = where;
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildDeleteQuery(this.config);
    }
    toSQL() {
        const { typings: _typings, ...rest } = this.dialect.sqlToQuery(this.getSQL());
        return rest;
    }
    prepare() {
        return this.session.prepareQuery(this.dialect.sqlToQuery(this.getSQL()), this.config.returning);
    }
}
_a$9 = alias.entityKind;
MySqlDelete[_a$9] = 'MySqlDelete';

var _a$8, _b$5;
class MySqlInsertBuilder {
    constructor(table, session, dialect) {
        this.table = table;
        this.session = session;
        this.dialect = dialect;
        this.shouldIgnore = false;
    }
    ignore() {
        this.shouldIgnore = true;
        return this;
    }
    values(values) {
        values = Array.isArray(values) ? values : [values];
        if (values.length === 0) {
            throw new Error('values() must be called with at least one value');
        }
        const mappedValues = values.map((entry) => {
            const result = {};
            const cols = this.table[alias.Table.Symbol.Columns];
            for (const colKey of Object.keys(entry)) {
                const colValue = entry[colKey];
                result[colKey] = alias.is(colValue, alias.SQL) ? colValue : new alias.Param(colValue, cols[colKey]);
            }
            return result;
        });
        return new MySqlInsert(this.table, mappedValues, this.shouldIgnore, this.session, this.dialect);
    }
}
_a$8 = alias.entityKind;
MySqlInsertBuilder[_a$8] = 'MySqlInsertBuilder';
class MySqlInsert extends alias.QueryPromise {
    constructor(table, values, ignore, session, dialect) {
        super();
        this.session = session;
        this.dialect = dialect;
        this.execute = (placeholderValues) => {
            return this.prepare().execute(placeholderValues);
        };
        this.createIterator = () => {
            const self = this;
            return async function* (placeholderValues) {
                yield* self.prepare().iterator(placeholderValues);
            };
        };
        this.iterator = this.createIterator();
        this.config = { table, values, ignore };
    }
    onDuplicateKeyUpdate(config) {
        const setSql = this.dialect.buildUpdateSet(this.config.table, alias.mapUpdateSet(this.config.table, config.set));
        this.config.onConflict = alias.sql `update ${setSql}`;
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildInsertQuery(this.config);
    }
    toSQL() {
        const { typings: _typings, ...rest } = this.dialect.sqlToQuery(this.getSQL());
        return rest;
    }
    prepare() {
        return this.session.prepareQuery(this.dialect.sqlToQuery(this.getSQL()), undefined);
    }
}
_b$5 = alias.entityKind;
MySqlInsert[_b$5] = 'MySqlInsert';

var _a$7, _b$4, _c$2, _d, _e;
class ViewBuilderCore {
    constructor(name, schema) {
        this.name = name;
        this.schema = schema;
        this.config = {};
    }
    algorithm(algorithm) {
        this.config.algorithm = algorithm;
        return this;
    }
    definer(definer) {
        this.config.definer = definer;
        return this;
    }
    sqlSecurity(sqlSecurity) {
        this.config.sqlSecurity = sqlSecurity;
        return this;
    }
    withCheckOption(withCheckOption) {
        this.config.withCheckOption = withCheckOption ?? 'cascaded';
        return this;
    }
}
_a$7 = alias.entityKind;
ViewBuilderCore[_a$7] = 'MySqlViewBuilder';
class ViewBuilder extends ViewBuilderCore {
    as(qb) {
        if (typeof qb === 'function') {
            qb = qb(new QueryBuilder());
        }
        const selectionProxy = new alias.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        });
        const aliasedSelection = new Proxy(qb.getSelectedFields(), selectionProxy);
        return new Proxy(new MySqlView({
            mysqlConfig: this.config,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: aliasedSelection,
                query: qb.getSQL().inlineParams(),
            },
        }), selectionProxy);
    }
}
_b$4 = alias.entityKind;
ViewBuilder[_b$4] = 'MySqlViewBuilder';
class ManualViewBuilder extends ViewBuilderCore {
    constructor(name, columns, schema) {
        super(name, schema);
        this.columns = alias.getTableColumns(mysqlTable(name, columns));
    }
    existing() {
        return new Proxy(new MySqlView({
            mysqlConfig: undefined,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: this.columns,
                query: undefined,
            },
        }), new alias.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
    as(query) {
        return new Proxy(new MySqlView({
            mysqlConfig: this.config,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: this.columns,
                query: query.inlineParams(),
            },
        }), new alias.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
}
_c$2 = alias.entityKind;
ManualViewBuilder[_c$2] = 'MySqlManualViewBuilder';
class MySqlViewBase extends alias.View {
}
_d = alias.entityKind;
MySqlViewBase[_d] = 'MySqlViewBase';
const MySqlViewConfig = Symbol.for('drizzle:MySqlViewConfig');
class MySqlView extends MySqlViewBase {
    constructor({ mysqlConfig, config }) {
        super(config);
        this[MySqlViewConfig] = mysqlConfig;
    }
}
_e = alias.entityKind;
MySqlView[_e] = 'MySqlView';
/** @internal */
function mysqlViewWithSchema(name, selection, schema) {
    if (selection) {
        return new ManualViewBuilder(name, selection, schema);
    }
    return new ViewBuilder(name, schema);
}
function mysqlView(name, selection) {
    return mysqlViewWithSchema(name, selection, undefined);
}

var _a$6;
// TODO find out how to use all/values. Seems like I need those functions
// Build project
// copy runtime tests to be sure it's working
// Add mysql to drizzle-kit
// Add Planetscale Driver and create example repo
class MySqlDialect {
    async migrate(migrations, session, config) {
        const migrationsTable = config.migrationsTable ?? '__drizzle_migrations';
        const migrationTableCreate = alias.sql `
			create table if not exists ${alias.sql.identifier(migrationsTable)} (
				id serial primary key,
				hash text not null,
				created_at bigint
			)
		`;
        await session.execute(migrationTableCreate);
        const dbMigrations = await session.all(alias.sql `select id, hash, created_at from ${alias.sql.identifier(migrationsTable)} order by created_at desc limit 1`);
        const lastDbMigration = dbMigrations[0];
        await session.transaction(async (tx) => {
            for (const migration of migrations) {
                if (!lastDbMigration
                    || Number(lastDbMigration.created_at) < migration.folderMillis) {
                    for (const stmt of migration.sql) {
                        await tx.execute(alias.sql.raw(stmt));
                    }
                    await tx.execute(alias.sql `insert into ${alias.sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`) values(${migration.hash}, ${migration.folderMillis})`);
                }
            }
        });
    }
    escapeName(name) {
        return `\`${name}\``;
    }
    escapeParam(_num) {
        return `?`;
    }
    escapeString(str) {
        return `'${str.replace(/'/g, "''")}'`;
    }
    buildDeleteQuery({ table, where, returning }) {
        const returningSql = returning
            ? alias.sql ` returning ${this.buildSelection(returning, { isSingleTable: true })}`
            : undefined;
        const whereSql = where ? alias.sql ` where ${where}` : undefined;
        return alias.sql `delete from ${table}${whereSql}${returningSql}`;
    }
    buildUpdateSet(table, set) {
        const setEntries = Object.entries(set);
        const setSize = setEntries.length;
        return alias.sql.join(setEntries
            .flatMap(([colName, value], i) => {
            const col = table[alias.Table.Symbol.Columns][colName];
            const res = alias.sql `${alias.sql.identifier(col.name)} = ${value}`;
            if (i < setSize - 1) {
                return [res, alias.sql.raw(', ')];
            }
            return [res];
        }));
    }
    buildUpdateQuery({ table, set, where, returning }) {
        const setSql = this.buildUpdateSet(table, set);
        const returningSql = returning
            ? alias.sql ` returning ${this.buildSelection(returning, { isSingleTable: true })}`
            : undefined;
        const whereSql = where ? alias.sql ` where ${where}` : undefined;
        return alias.sql `update ${table} set ${setSql}${whereSql}${returningSql}`;
    }
    /**
     * Builds selection SQL with provided fields/expressions
     *
     * Examples:
     *
     * `select <selection> from`
     *
     * `insert ... returning <selection>`
     *
     * If `isSingleTable` is true, then columns won't be prefixed with table name
     */
    buildSelection(fields, { isSingleTable = false } = {}) {
        const columnsLen = fields.length;
        const chunks = fields
            .flatMap(({ field }, i) => {
            const chunk = [];
            if (alias.is(field, alias.SQL.Aliased) && field.isSelectionField) {
                chunk.push(alias.sql.identifier(field.fieldAlias));
            }
            else if (alias.is(field, alias.SQL.Aliased) || alias.is(field, alias.SQL)) {
                const query = alias.is(field, alias.SQL.Aliased) ? field.sql : field;
                if (isSingleTable) {
                    chunk.push(new alias.SQL(query.queryChunks.map((c) => {
                        if (alias.is(c, MySqlColumn)) {
                            return alias.sql.identifier(c.name);
                        }
                        return c;
                    })));
                }
                else {
                    chunk.push(query);
                }
                if (alias.is(field, alias.SQL.Aliased)) {
                    chunk.push(alias.sql ` as ${alias.sql.identifier(field.fieldAlias)}`);
                }
            }
            else if (alias.is(field, alias.Column)) {
                if (isSingleTable) {
                    chunk.push(alias.sql.identifier(field.name));
                }
                else {
                    chunk.push(field);
                }
            }
            if (i < columnsLen - 1) {
                chunk.push(alias.sql `, `);
            }
            return chunk;
        });
        return alias.sql.join(chunks);
    }
    buildSelectQuery({ withList, fields, fieldsFlat, where, having, table, joins, orderBy, groupBy, limit, offset, lockingClause, distinct, }) {
        const fieldsList = fieldsFlat ?? alias.orderSelectedFields(fields);
        for (const f of fieldsList) {
            if (alias.is(f.field, alias.Column)
                && alias.getTableName(f.field.table)
                    !== (alias.is(table, alias.Subquery)
                        ? table[alias.SubqueryConfig].alias
                        : alias.is(table, MySqlViewBase)
                            ? table[alias.ViewBaseConfig].name
                            : alias.is(table, alias.SQL)
                                ? undefined
                                : alias.getTableName(table))
                && !((table) => joins?.some(({ alias: alias$1 }) => alias$1 === (table[alias.Table.Symbol.IsAlias] ? alias.getTableName(table) : table[alias.Table.Symbol.BaseName])))(f.field.table)) {
                const tableName = alias.getTableName(f.field.table);
                throw new Error(`Your "${f.path.join('->')}" field references a column "${tableName}"."${f.field.name}", but the table "${tableName}" is not part of the query! Did you forget to join it?`);
            }
        }
        const isSingleTable = !joins || joins.length === 0;
        let withSql;
        if (withList?.length) {
            const withSqlChunks = [alias.sql `with `];
            for (const [i, w] of withList.entries()) {
                withSqlChunks.push(alias.sql `${alias.sql.identifier(w[alias.SubqueryConfig].alias)} as (${w[alias.SubqueryConfig].sql})`);
                if (i < withList.length - 1) {
                    withSqlChunks.push(alias.sql `, `);
                }
            }
            withSqlChunks.push(alias.sql ` `);
            withSql = alias.sql.join(withSqlChunks);
        }
        const distinctSql = distinct ? alias.sql ` distinct` : undefined;
        const selection = this.buildSelection(fieldsList, { isSingleTable });
        const tableSql = (() => {
            if (alias.is(table, alias.Table) && table[alias.Table.Symbol.OriginalName] !== table[alias.Table.Symbol.Name]) {
                return alias.sql `${alias.sql.identifier(table[alias.Table.Symbol.OriginalName])} ${alias.sql.identifier(table[alias.Table.Symbol.Name])}`;
            }
            return table;
        })();
        const joinsArray = [];
        if (joins) {
            for (const [index, joinMeta] of joins.entries()) {
                if (index === 0) {
                    joinsArray.push(alias.sql ` `);
                }
                const table = joinMeta.table;
                const lateralSql = joinMeta.lateral ? alias.sql ` lateral` : undefined;
                if (alias.is(table, MySqlTable)) {
                    const tableName = table[MySqlTable.Symbol.Name];
                    const tableSchema = table[MySqlTable.Symbol.Schema];
                    const origTableName = table[MySqlTable.Symbol.OriginalName];
                    const alias$1 = tableName === origTableName ? undefined : joinMeta.alias;
                    joinsArray.push(alias.sql `${alias.sql.raw(joinMeta.joinType)} join${lateralSql} ${tableSchema ? alias.sql `${alias.sql.identifier(tableSchema)}.` : undefined}${alias.sql.identifier(origTableName)}${alias$1 && alias.sql ` ${alias.sql.identifier(alias$1)}`} on ${joinMeta.on}`);
                }
                else if (alias.is(table, alias.View)) {
                    const viewName = table[alias.ViewBaseConfig].name;
                    const viewSchema = table[alias.ViewBaseConfig].schema;
                    const origViewName = table[alias.ViewBaseConfig].originalName;
                    const alias$1 = viewName === origViewName ? undefined : joinMeta.alias;
                    joinsArray.push(alias.sql `${alias.sql.raw(joinMeta.joinType)} join${lateralSql} ${viewSchema ? alias.sql `${alias.sql.identifier(viewSchema)}.` : undefined}${alias.sql.identifier(origViewName)}${alias$1 && alias.sql ` ${alias.sql.identifier(alias$1)}`} on ${joinMeta.on}`);
                }
                else {
                    joinsArray.push(alias.sql `${alias.sql.raw(joinMeta.joinType)} join${lateralSql} ${table} on ${joinMeta.on}`);
                }
                if (index < joins.length - 1) {
                    joinsArray.push(alias.sql ` `);
                }
            }
        }
        const joinsSql = alias.sql.join(joinsArray);
        const whereSql = where ? alias.sql ` where ${where}` : undefined;
        const havingSql = having ? alias.sql ` having ${having}` : undefined;
        let orderBySql;
        if (orderBy && orderBy.length > 0) {
            orderBySql = alias.sql ` order by ${alias.sql.join(orderBy, alias.sql `, `)}`;
        }
        let groupBySql;
        if (groupBy && groupBy.length > 0) {
            groupBySql = alias.sql ` group by ${alias.sql.join(groupBy, alias.sql `, `)}`;
        }
        const limitSql = limit ? alias.sql ` limit ${limit}` : undefined;
        const offsetSql = offset ? alias.sql ` offset ${offset}` : undefined;
        let lockingClausesSql;
        if (lockingClause) {
            const { config, strength } = lockingClause;
            lockingClausesSql = alias.sql ` for ${alias.sql.raw(strength)}`;
            if (config.noWait) {
                lockingClausesSql.append(alias.sql ` no wait`);
            }
            else if (config.skipLocked) {
                lockingClausesSql.append(alias.sql ` skip locked`);
            }
        }
        return alias.sql `${withSql}select${distinctSql} ${selection} from ${tableSql}${joinsSql}${whereSql}${groupBySql}${havingSql}${orderBySql}${limitSql}${offsetSql}${lockingClausesSql}`;
    }
    buildInsertQuery({ table, values, ignore, onConflict }) {
        const isSingleValue = values.length === 1;
        const valuesSqlList = [];
        const columns = table[alias.Table.Symbol.Columns];
        const colEntries = isSingleValue
            ? Object.keys(values[0]).map((fieldName) => [fieldName, columns[fieldName]])
            : Object.entries(columns);
        const insertOrder = colEntries.map(([, column]) => alias.sql.identifier(column.name));
        for (const [valueIndex, value] of values.entries()) {
            const valueList = [];
            for (const [fieldName] of colEntries) {
                const colValue = value[fieldName];
                if (colValue === undefined || (alias.is(colValue, alias.Param) && colValue.value === undefined)) {
                    valueList.push(alias.sql `default`);
                }
                else {
                    valueList.push(colValue);
                }
            }
            valuesSqlList.push(valueList);
            if (valueIndex < values.length - 1) {
                valuesSqlList.push(alias.sql `, `);
            }
        }
        const valuesSql = alias.sql.join(valuesSqlList);
        const ignoreSql = ignore ? alias.sql ` ignore` : undefined;
        const onConflictSql = onConflict ? alias.sql ` on duplicate key ${onConflict}` : undefined;
        return alias.sql `insert${ignoreSql} into ${table} ${insertOrder} values ${valuesSql}${onConflictSql}`;
    }
    sqlToQuery(sql) {
        return sql.toQuery({
            escapeName: this.escapeName,
            escapeParam: this.escapeParam,
            escapeString: this.escapeString,
        });
    }
    buildRelationalQuery({ fullSchema, schema, tableNamesMap, table, tableConfig, queryConfig: config, tableAlias, nestedQueryRelation, joinOn, }) {
        let selection = [];
        let limit, offset, orderBy, where;
        const joins = [];
        if (config === true) {
            const selectionEntries = Object.entries(tableConfig.columns);
            selection = selectionEntries.map(([key, value]) => ({
                dbKey: value.name,
                tsKey: key,
                field: alias.aliasedTableColumn(value, tableAlias),
                relationTableTsKey: undefined,
                isJson: false,
                selection: [],
            }));
        }
        else {
            const aliasedColumns = Object.fromEntries(Object.entries(tableConfig.columns).map(([key, value]) => [key, alias.aliasedTableColumn(value, tableAlias)]));
            if (config.where) {
                const whereSql = typeof config.where === 'function'
                    ? config.where(aliasedColumns, alias.getOperators())
                    : config.where;
                where = whereSql && alias.mapColumnsInSQLToAlias(whereSql, tableAlias);
            }
            const fieldsSelection = [];
            let selectedColumns = [];
            // Figure out which columns to select
            if (config.columns) {
                let isIncludeMode = false;
                for (const [field, value] of Object.entries(config.columns)) {
                    if (value === undefined) {
                        continue;
                    }
                    if (field in tableConfig.columns) {
                        if (!isIncludeMode && value === true) {
                            isIncludeMode = true;
                        }
                        selectedColumns.push(field);
                    }
                }
                if (selectedColumns.length > 0) {
                    selectedColumns = isIncludeMode
                        ? selectedColumns.filter((c) => config.columns?.[c] === true)
                        : Object.keys(tableConfig.columns).filter((key) => !selectedColumns.includes(key));
                }
            }
            else {
                // Select all columns if selection is not specified
                selectedColumns = Object.keys(tableConfig.columns);
            }
            for (const field of selectedColumns) {
                const column = tableConfig.columns[field];
                fieldsSelection.push({ tsKey: field, value: column });
            }
            let selectedRelations = [];
            // Figure out which relations to select
            if (config.with) {
                selectedRelations = Object.entries(config.with)
                    .filter((entry) => !!entry[1])
                    .map(([tsKey, queryConfig]) => ({ tsKey, queryConfig, relation: tableConfig.relations[tsKey] }));
            }
            let extras;
            // Figure out which extras to select
            if (config.extras) {
                extras = typeof config.extras === 'function'
                    ? config.extras(aliasedColumns, { sql: alias.sql })
                    : config.extras;
                for (const [tsKey, value] of Object.entries(extras)) {
                    fieldsSelection.push({
                        tsKey,
                        value: alias.mapColumnsInAliasedSQLToAlias(value, tableAlias),
                    });
                }
            }
            // Transform `fieldsSelection` into `selection`
            // `fieldsSelection` shouldn't be used after this point
            for (const { tsKey, value } of fieldsSelection) {
                selection.push({
                    dbKey: alias.is(value, alias.SQL.Aliased) ? value.fieldAlias : tableConfig.columns[tsKey].name,
                    tsKey,
                    field: alias.is(value, alias.Column) ? alias.aliasedTableColumn(value, tableAlias) : value,
                    relationTableTsKey: undefined,
                    isJson: false,
                    selection: [],
                });
            }
            let orderByOrig = typeof config.orderBy === 'function'
                ? config.orderBy(aliasedColumns, alias.getOrderByOperators())
                : config.orderBy ?? [];
            if (!Array.isArray(orderByOrig)) {
                orderByOrig = [orderByOrig];
            }
            orderBy = orderByOrig.map((orderByValue) => {
                if (alias.is(orderByValue, alias.Column)) {
                    return alias.aliasedTableColumn(orderByValue, tableAlias);
                }
                return alias.mapColumnsInSQLToAlias(orderByValue, tableAlias);
            });
            limit = config.limit;
            offset = config.offset;
            // Process all relations
            for (const { tsKey: selectedRelationTsKey, queryConfig: selectedRelationConfigValue, relation, } of selectedRelations) {
                const normalizedRelation = alias.normalizeRelation(schema, tableNamesMap, relation);
                const relationTableName = relation.referencedTable[alias.Table.Symbol.Name];
                const relationTableTsName = tableNamesMap[relationTableName];
                const relationTableAlias = `${tableAlias}_${selectedRelationTsKey}`;
                const joinOn = alias.and(...normalizedRelation.fields.map((field, i) => alias.eq(alias.aliasedTableColumn(normalizedRelation.references[i], relationTableAlias), alias.aliasedTableColumn(field, tableAlias))));
                const builtRelation = this.buildRelationalQuery({
                    fullSchema,
                    schema,
                    tableNamesMap,
                    table: fullSchema[relationTableTsName],
                    tableConfig: schema[relationTableTsName],
                    queryConfig: alias.is(relation, alias.One)
                        ? (selectedRelationConfigValue === true
                            ? { limit: 1 }
                            : { ...selectedRelationConfigValue, limit: 1 })
                        : selectedRelationConfigValue,
                    tableAlias: relationTableAlias,
                    joinOn,
                    nestedQueryRelation: relation,
                });
                const field = alias.sql `${alias.sql.identifier(relationTableAlias)}.${alias.sql.identifier('data')}`.as(selectedRelationTsKey);
                joins.push({
                    on: alias.sql `true`,
                    table: new alias.Subquery(builtRelation.sql, {}, relationTableAlias),
                    alias: relationTableAlias,
                    joinType: 'left',
                    lateral: true,
                });
                selection.push({
                    dbKey: selectedRelationTsKey,
                    tsKey: selectedRelationTsKey,
                    field,
                    relationTableTsKey: relationTableTsName,
                    isJson: true,
                    selection: builtRelation.selection,
                });
            }
        }
        if (selection.length === 0) {
            throw new alias.DrizzleError(`No fields selected for table "${tableConfig.tsName}" ("${tableAlias}")`);
        }
        let result;
        where = alias.and(joinOn, where);
        if (nestedQueryRelation) {
            let field = alias.sql `json_array(${alias.sql.join(selection.map(({ field, tsKey, isJson }) => isJson
                ? alias.sql `${alias.sql.identifier(`${tableAlias}_${tsKey}`)}.${alias.sql.identifier('data')}`
                : alias.is(field, alias.SQL.Aliased)
                    ? field.sql
                    : field), alias.sql `, `)})`;
            if (alias.is(nestedQueryRelation, alias.Many)) {
                field = alias.sql `coalesce(json_arrayagg(${field}), json_array())`;
            }
            const nestedSelection = [{
                    dbKey: 'data',
                    tsKey: 'data',
                    field: field.as('data'),
                    isJson: true,
                    relationTableTsKey: tableConfig.tsName,
                    selection,
                }];
            const needsSubquery = limit !== undefined || offset !== undefined || (orderBy?.length ?? 0) > 0;
            if (needsSubquery) {
                result = this.buildSelectQuery({
                    table: alias.aliasedTable(table, tableAlias),
                    fields: {},
                    fieldsFlat: [
                        {
                            path: [],
                            field: alias.sql.raw('*'),
                        },
                        ...(((orderBy?.length ?? 0) > 0)
                            ? [{
                                    path: [],
                                    field: alias.sql `row_number() over (order by ${alias.sql.join(orderBy, alias.sql `, `)})`,
                                }]
                            : []),
                    ],
                    where,
                    limit,
                    offset,
                });
                where = undefined;
                limit = undefined;
                offset = undefined;
                orderBy = undefined;
            }
            else {
                result = alias.aliasedTable(table, tableAlias);
            }
            result = this.buildSelectQuery({
                table: alias.is(result, MySqlTable) ? result : new alias.Subquery(result, {}, tableAlias),
                fields: {},
                fieldsFlat: nestedSelection.map(({ field }) => ({
                    path: [],
                    field: alias.is(field, alias.Column) ? alias.aliasedTableColumn(field, tableAlias) : field,
                })),
                joins,
                where,
                limit,
                offset,
                orderBy,
            });
        }
        else {
            result = this.buildSelectQuery({
                table: alias.aliasedTable(table, tableAlias),
                fields: {},
                fieldsFlat: selection.map(({ field }) => ({
                    path: [],
                    field: alias.is(field, alias.Column) ? alias.aliasedTableColumn(field, tableAlias) : field,
                })),
                joins,
                where,
                limit,
                offset,
                orderBy,
            });
        }
        return {
            tableTsKey: tableConfig.tsName,
            sql: result,
            selection,
        };
    }
    buildRelationalQueryWithoutLateralSubqueries({ fullSchema, schema, tableNamesMap, table, tableConfig, queryConfig: config, tableAlias, nestedQueryRelation, joinOn, }) {
        let selection = [];
        let limit, offset, orderBy = [], where;
        if (config === true) {
            const selectionEntries = Object.entries(tableConfig.columns);
            selection = selectionEntries.map(([key, value]) => ({
                dbKey: value.name,
                tsKey: key,
                field: alias.aliasedTableColumn(value, tableAlias),
                relationTableTsKey: undefined,
                isJson: false,
                selection: [],
            }));
        }
        else {
            const aliasedColumns = Object.fromEntries(Object.entries(tableConfig.columns).map(([key, value]) => [key, alias.aliasedTableColumn(value, tableAlias)]));
            if (config.where) {
                const whereSql = typeof config.where === 'function'
                    ? config.where(aliasedColumns, alias.getOperators())
                    : config.where;
                where = whereSql && alias.mapColumnsInSQLToAlias(whereSql, tableAlias);
            }
            const fieldsSelection = [];
            let selectedColumns = [];
            // Figure out which columns to select
            if (config.columns) {
                let isIncludeMode = false;
                for (const [field, value] of Object.entries(config.columns)) {
                    if (value === undefined) {
                        continue;
                    }
                    if (field in tableConfig.columns) {
                        if (!isIncludeMode && value === true) {
                            isIncludeMode = true;
                        }
                        selectedColumns.push(field);
                    }
                }
                if (selectedColumns.length > 0) {
                    selectedColumns = isIncludeMode
                        ? selectedColumns.filter((c) => config.columns?.[c] === true)
                        : Object.keys(tableConfig.columns).filter((key) => !selectedColumns.includes(key));
                }
            }
            else {
                // Select all columns if selection is not specified
                selectedColumns = Object.keys(tableConfig.columns);
            }
            for (const field of selectedColumns) {
                const column = tableConfig.columns[field];
                fieldsSelection.push({ tsKey: field, value: column });
            }
            let selectedRelations = [];
            // Figure out which relations to select
            if (config.with) {
                selectedRelations = Object.entries(config.with)
                    .filter((entry) => !!entry[1])
                    .map(([tsKey, queryConfig]) => ({ tsKey, queryConfig, relation: tableConfig.relations[tsKey] }));
            }
            let extras;
            // Figure out which extras to select
            if (config.extras) {
                extras = typeof config.extras === 'function'
                    ? config.extras(aliasedColumns, { sql: alias.sql })
                    : config.extras;
                for (const [tsKey, value] of Object.entries(extras)) {
                    fieldsSelection.push({
                        tsKey,
                        value: alias.mapColumnsInAliasedSQLToAlias(value, tableAlias),
                    });
                }
            }
            // Transform `fieldsSelection` into `selection`
            // `fieldsSelection` shouldn't be used after this point
            for (const { tsKey, value } of fieldsSelection) {
                selection.push({
                    dbKey: alias.is(value, alias.SQL.Aliased) ? value.fieldAlias : tableConfig.columns[tsKey].name,
                    tsKey,
                    field: alias.is(value, alias.Column) ? alias.aliasedTableColumn(value, tableAlias) : value,
                    relationTableTsKey: undefined,
                    isJson: false,
                    selection: [],
                });
            }
            let orderByOrig = typeof config.orderBy === 'function'
                ? config.orderBy(aliasedColumns, alias.getOrderByOperators())
                : config.orderBy ?? [];
            if (!Array.isArray(orderByOrig)) {
                orderByOrig = [orderByOrig];
            }
            orderBy = orderByOrig.map((orderByValue) => {
                if (alias.is(orderByValue, alias.Column)) {
                    return alias.aliasedTableColumn(orderByValue, tableAlias);
                }
                return alias.mapColumnsInSQLToAlias(orderByValue, tableAlias);
            });
            limit = config.limit;
            offset = config.offset;
            // Process all relations
            for (const { tsKey: selectedRelationTsKey, queryConfig: selectedRelationConfigValue, relation, } of selectedRelations) {
                const normalizedRelation = alias.normalizeRelation(schema, tableNamesMap, relation);
                const relationTableName = relation.referencedTable[alias.Table.Symbol.Name];
                const relationTableTsName = tableNamesMap[relationTableName];
                const relationTableAlias = `${tableAlias}_${selectedRelationTsKey}`;
                const joinOn = alias.and(...normalizedRelation.fields.map((field, i) => alias.eq(alias.aliasedTableColumn(normalizedRelation.references[i], relationTableAlias), alias.aliasedTableColumn(field, tableAlias))));
                const builtRelation = this.buildRelationalQueryWithoutLateralSubqueries({
                    fullSchema,
                    schema,
                    tableNamesMap,
                    table: fullSchema[relationTableTsName],
                    tableConfig: schema[relationTableTsName],
                    queryConfig: alias.is(relation, alias.One)
                        ? (selectedRelationConfigValue === true
                            ? { limit: 1 }
                            : { ...selectedRelationConfigValue, limit: 1 })
                        : selectedRelationConfigValue,
                    tableAlias: relationTableAlias,
                    joinOn,
                    nestedQueryRelation: relation,
                });
                let fieldSql = alias.sql `(${builtRelation.sql})`;
                if (alias.is(relation, alias.Many)) {
                    fieldSql = alias.sql `coalesce(${fieldSql}, json_array())`;
                }
                const field = fieldSql.as(selectedRelationTsKey);
                selection.push({
                    dbKey: selectedRelationTsKey,
                    tsKey: selectedRelationTsKey,
                    field,
                    relationTableTsKey: relationTableTsName,
                    isJson: true,
                    selection: builtRelation.selection,
                });
            }
        }
        if (selection.length === 0) {
            throw new alias.DrizzleError(`No fields selected for table "${tableConfig.tsName}" ("${tableAlias}"). You need to have at least one item in "columns", "with" or "extras". If you need to select all columns, omit the "columns" key or set it to undefined.`);
        }
        let result;
        where = alias.and(joinOn, where);
        if (nestedQueryRelation) {
            let field = alias.sql `json_array(${alias.sql.join(selection.map(({ field }) => alias.is(field, MySqlColumn) ? alias.sql.identifier(field.name) : alias.is(field, alias.SQL.Aliased) ? field.sql : field), alias.sql `, `)})`;
            if (alias.is(nestedQueryRelation, alias.Many)) {
                field = alias.sql `json_arrayagg(${field})`;
            }
            const nestedSelection = [{
                    dbKey: 'data',
                    tsKey: 'data',
                    field,
                    isJson: true,
                    relationTableTsKey: tableConfig.tsName,
                    selection,
                }];
            const needsSubquery = limit !== undefined || offset !== undefined || orderBy.length > 0;
            if (needsSubquery) {
                result = this.buildSelectQuery({
                    table: alias.aliasedTable(table, tableAlias),
                    fields: {},
                    fieldsFlat: [
                        {
                            path: [],
                            field: alias.sql.raw('*'),
                        },
                        ...(orderBy.length > 0)
                            ? [{
                                    path: [],
                                    field: alias.sql `row_number() over (order by ${alias.sql.join(orderBy, alias.sql `, `)})`,
                                }]
                            : [],
                    ],
                    where,
                    limit,
                    offset,
                });
                where = undefined;
                limit = undefined;
                offset = undefined;
                orderBy = undefined;
            }
            else {
                result = alias.aliasedTable(table, tableAlias);
            }
            result = this.buildSelectQuery({
                table: alias.is(result, MySqlTable) ? result : new alias.Subquery(result, {}, tableAlias),
                fields: {},
                fieldsFlat: nestedSelection.map(({ field }) => ({
                    path: [],
                    field: alias.is(field, alias.Column) ? alias.aliasedTableColumn(field, tableAlias) : field,
                })),
                where,
                limit,
                offset,
                orderBy,
            });
        }
        else {
            result = this.buildSelectQuery({
                table: alias.aliasedTable(table, tableAlias),
                fields: {},
                fieldsFlat: selection.map(({ field }) => ({
                    path: [],
                    field: alias.is(field, alias.Column) ? alias.aliasedTableColumn(field, tableAlias) : field,
                })),
                where,
                limit,
                offset,
                orderBy,
            });
        }
        return {
            tableTsKey: tableConfig.tsName,
            sql: result,
            selection,
        };
    }
}
_a$6 = alias.entityKind;
MySqlDialect[_a$6] = 'MySqlDialect';

var _a$5, _b$3, _c$1;
class MySqlSelectBuilder {
    constructor(config) {
        this.withList = [];
        this.fields = config.fields;
        this.session = config.session;
        this.dialect = config.dialect;
        if (config.withList) {
            this.withList = config.withList;
        }
        this.distinct = config.distinct;
    }
    from(source) {
        const isPartialSelect = !!this.fields;
        let fields;
        if (this.fields) {
            fields = this.fields;
        }
        else if (alias.is(source, alias.Subquery)) {
            // This is required to use the proxy handler to get the correct field values from the subquery
            fields = Object.fromEntries(Object.keys(source[alias.SubqueryConfig].selection).map((key) => [key, source[key]]));
        }
        else if (alias.is(source, MySqlViewBase)) {
            fields = source[alias.ViewBaseConfig].selectedFields;
        }
        else if (alias.is(source, alias.SQL)) {
            fields = {};
        }
        else {
            fields = alias.getTableColumns(source);
        }
        return new MySqlSelect({
            table: source,
            fields,
            isPartialSelect,
            session: this.session,
            dialect: this.dialect,
            withList: this.withList,
            distinct: this.distinct,
        });
    }
}
_a$5 = alias.entityKind;
MySqlSelectBuilder[_a$5] = 'MySqlSelectBuilder';
class MySqlSelectQueryBuilder extends alias.TypedQueryBuilder {
    constructor({ table, fields, isPartialSelect, session, dialect, withList, distinct }) {
        super();
        this.leftJoin = this.createJoin('left');
        this.rightJoin = this.createJoin('right');
        this.innerJoin = this.createJoin('inner');
        this.fullJoin = this.createJoin('full');
        this.config = {
            withList,
            table,
            fields: { ...fields },
            distinct,
        };
        this.isPartialSelect = isPartialSelect;
        this.session = session;
        this.dialect = dialect;
        this._ = {
            selectedFields: fields,
        };
        this.tableName = alias.getTableLikeName(table);
        this.joinsNotNullableMap = typeof this.tableName === 'string' ? { [this.tableName]: true } : {};
    }
    createJoin(joinType) {
        return (table, on) => {
            const baseTableName = this.tableName;
            const tableName = alias.getTableLikeName(table);
            if (typeof tableName === 'string' && this.config.joins?.some((join) => join.alias === tableName)) {
                throw new Error(`Alias "${tableName}" is already used in this query`);
            }
            if (!this.isPartialSelect) {
                // If this is the first join and this is not a partial select and we're not selecting from raw SQL, "move" the fields from the main table to the nested object
                if (Object.keys(this.joinsNotNullableMap).length === 1 && typeof baseTableName === 'string') {
                    this.config.fields = {
                        [baseTableName]: this.config.fields,
                    };
                }
                if (typeof tableName === 'string' && !alias.is(table, alias.SQL)) {
                    const selection = alias.is(table, alias.Subquery)
                        ? table[alias.SubqueryConfig].selection
                        : alias.is(table, alias.View)
                            ? table[alias.ViewBaseConfig].selectedFields
                            : table[alias.Table.Symbol.Columns];
                    this.config.fields[tableName] = selection;
                }
            }
            if (typeof on === 'function') {
                on = on(new Proxy(this.config.fields, new alias.SelectionProxyHandler({ sqlAliasedBehavior: 'sql', sqlBehavior: 'sql' })));
            }
            if (!this.config.joins) {
                this.config.joins = [];
            }
            this.config.joins.push({ on, table, joinType, alias: tableName });
            if (typeof tableName === 'string') {
                switch (joinType) {
                    case 'left': {
                        this.joinsNotNullableMap[tableName] = false;
                        break;
                    }
                    case 'right': {
                        this.joinsNotNullableMap = Object.fromEntries(Object.entries(this.joinsNotNullableMap).map(([key]) => [key, false]));
                        this.joinsNotNullableMap[tableName] = true;
                        break;
                    }
                    case 'inner': {
                        this.joinsNotNullableMap[tableName] = true;
                        break;
                    }
                    case 'full': {
                        this.joinsNotNullableMap = Object.fromEntries(Object.entries(this.joinsNotNullableMap).map(([key]) => [key, false]));
                        this.joinsNotNullableMap[tableName] = false;
                        break;
                    }
                }
            }
            return this;
        };
    }
    where(where) {
        if (typeof where === 'function') {
            where = where(new Proxy(this.config.fields, new alias.SelectionProxyHandler({ sqlAliasedBehavior: 'sql', sqlBehavior: 'sql' })));
        }
        this.config.where = where;
        return this;
    }
    having(having) {
        if (typeof having === 'function') {
            having = having(new Proxy(this.config.fields, new alias.SelectionProxyHandler({ sqlAliasedBehavior: 'sql', sqlBehavior: 'sql' })));
        }
        this.config.having = having;
        return this;
    }
    groupBy(...columns) {
        if (typeof columns[0] === 'function') {
            const groupBy = columns[0](new Proxy(this.config.fields, new alias.SelectionProxyHandler({ sqlAliasedBehavior: 'alias', sqlBehavior: 'sql' })));
            this.config.groupBy = Array.isArray(groupBy) ? groupBy : [groupBy];
        }
        else {
            this.config.groupBy = columns;
        }
        return this;
    }
    orderBy(...columns) {
        if (typeof columns[0] === 'function') {
            const orderBy = columns[0](new Proxy(this.config.fields, new alias.SelectionProxyHandler({ sqlAliasedBehavior: 'alias', sqlBehavior: 'sql' })));
            this.config.orderBy = Array.isArray(orderBy) ? orderBy : [orderBy];
        }
        else {
            this.config.orderBy = columns;
        }
        return this;
    }
    limit(limit) {
        this.config.limit = limit;
        return this;
    }
    offset(offset) {
        this.config.offset = offset;
        return this;
    }
    for(strength, config = {}) {
        this.config.lockingClause = { strength, config };
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildSelectQuery(this.config);
    }
    toSQL() {
        const { typings: _typings, ...rest } = this.dialect.sqlToQuery(this.getSQL());
        return rest;
    }
    as(alias$1) {
        return new Proxy(new alias.Subquery(this.getSQL(), this.config.fields, alias$1), new alias.SelectionProxyHandler({ alias: alias$1, sqlAliasedBehavior: 'alias', sqlBehavior: 'error' }));
    }
}
_b$3 = alias.entityKind;
MySqlSelectQueryBuilder[_b$3] = 'MySqlSelectQueryBuilder';
class MySqlSelect extends MySqlSelectQueryBuilder {
    constructor() {
        super(...arguments);
        this.execute = ((placeholderValues) => {
            return this.prepare().execute(placeholderValues);
        });
        this.createIterator = () => {
            const self = this;
            return async function* (placeholderValues) {
                yield* self.prepare().iterator(placeholderValues);
            };
        };
        this.iterator = this.createIterator();
    }
    prepare() {
        if (!this.session) {
            throw new Error('Cannot execute a query on a query builder. Please use a database instance instead.');
        }
        const fieldsList = alias.orderSelectedFields(this.config.fields);
        const query = this.session.prepareQuery(this.dialect.sqlToQuery(this.getSQL()), fieldsList);
        query.joinsNotNullableMap = this.joinsNotNullableMap;
        return query;
    }
}
_c$1 = alias.entityKind;
MySqlSelect[_c$1] = 'MySqlSelect';
alias.applyMixins(MySqlSelect, [alias.QueryPromise]);

var _a$4;
class QueryBuilder {
    $with(alias$1) {
        const queryBuilder = this;
        return {
            as(qb) {
                if (typeof qb === 'function') {
                    qb = qb(queryBuilder);
                }
                return new Proxy(new alias.WithSubquery(qb.getSQL(), qb.getSelectedFields(), alias$1, true), new alias.SelectionProxyHandler({ alias: alias$1, sqlAliasedBehavior: 'alias', sqlBehavior: 'error' }));
            },
        };
    }
    with(...queries) {
        const self = this;
        function select(fields) {
            return new MySqlSelectBuilder({
                fields: fields ?? undefined,
                session: undefined,
                dialect: self.getDialect(),
                withList: queries,
            });
        }
        function selectDistinct(fields) {
            return new MySqlSelectBuilder({
                fields: fields ?? undefined,
                session: undefined,
                dialect: self.getDialect(),
                withList: queries,
                distinct: true,
            });
        }
        return { select, selectDistinct };
    }
    select(fields) {
        return new MySqlSelectBuilder({ fields: fields ?? undefined, session: undefined, dialect: this.getDialect() });
    }
    selectDistinct(fields) {
        return new MySqlSelectBuilder({
            fields: fields ?? undefined,
            session: undefined,
            dialect: this.getDialect(),
            distinct: true,
        });
    }
    // Lazy load dialect to avoid circular dependency
    getDialect() {
        if (!this.dialect) {
            this.dialect = new MySqlDialect();
        }
        return this.dialect;
    }
}
_a$4 = alias.entityKind;
QueryBuilder[_a$4] = 'MySqlQueryBuilder';

var _a$3, _b$2;
class MySqlUpdateBuilder {
    constructor(table, session, dialect) {
        this.table = table;
        this.session = session;
        this.dialect = dialect;
    }
    set(values) {
        return new MySqlUpdate(this.table, alias.mapUpdateSet(this.table, values), this.session, this.dialect);
    }
}
_a$3 = alias.entityKind;
MySqlUpdateBuilder[_a$3] = 'MySqlUpdateBuilder';
class MySqlUpdate extends alias.QueryPromise {
    constructor(table, set, session, dialect) {
        super();
        this.session = session;
        this.dialect = dialect;
        this.execute = (placeholderValues) => {
            return this.prepare().execute(placeholderValues);
        };
        this.createIterator = () => {
            const self = this;
            return async function* (placeholderValues) {
                yield* self.prepare().iterator(placeholderValues);
            };
        };
        this.iterator = this.createIterator();
        this.config = { set, table };
    }
    where(where) {
        this.config.where = where;
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildUpdateQuery(this.config);
    }
    toSQL() {
        const { typings: _typings, ...rest } = this.dialect.sqlToQuery(this.getSQL());
        return rest;
    }
    prepare() {
        return this.session.prepareQuery(this.dialect.sqlToQuery(this.getSQL()), this.config.returning);
    }
}
_b$2 = alias.entityKind;
MySqlUpdate[_b$2] = 'MySqlUpdate';

var _a$2, _b$1;
class RelationalQueryBuilder {
    constructor(fullSchema, schema, tableNamesMap, table, tableConfig, dialect, session, mode) {
        this.fullSchema = fullSchema;
        this.schema = schema;
        this.tableNamesMap = tableNamesMap;
        this.table = table;
        this.tableConfig = tableConfig;
        this.dialect = dialect;
        this.session = session;
        this.mode = mode;
    }
    findMany(config) {
        return new MySqlRelationalQuery(this.fullSchema, this.schema, this.tableNamesMap, this.table, this.tableConfig, this.dialect, this.session, config ? config : {}, 'many', this.mode);
    }
    findFirst(config) {
        return new MySqlRelationalQuery(this.fullSchema, this.schema, this.tableNamesMap, this.table, this.tableConfig, this.dialect, this.session, config ? { ...config, limit: 1 } : { limit: 1 }, 'first', this.mode);
    }
}
_a$2 = alias.entityKind;
RelationalQueryBuilder[_a$2] = 'MySqlRelationalQueryBuilder';
class MySqlRelationalQuery extends alias.QueryPromise {
    constructor(fullSchema, schema, tableNamesMap, table, tableConfig, dialect, session, config, queryMode, mode) {
        super();
        this.fullSchema = fullSchema;
        this.schema = schema;
        this.tableNamesMap = tableNamesMap;
        this.table = table;
        this.tableConfig = tableConfig;
        this.dialect = dialect;
        this.session = session;
        this.config = config;
        this.queryMode = queryMode;
        this.mode = mode;
    }
    prepare() {
        const query = this.mode === 'planetscale'
            ? this.dialect.buildRelationalQueryWithoutLateralSubqueries({
                fullSchema: this.fullSchema,
                schema: this.schema,
                tableNamesMap: this.tableNamesMap,
                table: this.table,
                tableConfig: this.tableConfig,
                queryConfig: this.config,
                tableAlias: this.tableConfig.tsName,
            })
            : this.dialect.buildRelationalQuery({
                fullSchema: this.fullSchema,
                schema: this.schema,
                tableNamesMap: this.tableNamesMap,
                table: this.table,
                tableConfig: this.tableConfig,
                queryConfig: this.config,
                tableAlias: this.tableConfig.tsName,
            });
        const builtQuery = this.dialect.sqlToQuery(query.sql);
        return this.session.prepareQuery(builtQuery, undefined, (rawRows) => {
            const rows = rawRows.map((row) => alias.mapRelationalRow(this.schema, this.tableConfig, row, query.selection));
            if (this.queryMode === 'first') {
                return rows[0];
            }
            return rows;
        });
    }
    execute() {
        return this.prepare().execute();
    }
}
_b$1 = alias.entityKind;
MySqlRelationalQuery[_b$1] = 'MySqlRelationalQuery';

var _a$1;
class MySqlDatabase {
    constructor(
    /** @internal */
    dialect, 
    /** @internal */
    session, schema, mode) {
        this.dialect = dialect;
        this.session = session;
        this.mode = mode;
        this._ = schema
            ? { schema: schema.schema, tableNamesMap: schema.tableNamesMap }
            : { schema: undefined, tableNamesMap: {} };
        this.query = {};
        if (this._.schema) {
            for (const [tableName, columns] of Object.entries(this._.schema)) {
                this.query[tableName] =
                    new RelationalQueryBuilder(schema.fullSchema, this._.schema, this._.tableNamesMap, schema.fullSchema[tableName], columns, dialect, session, this.mode);
            }
        }
    }
    $with(alias$1) {
        return {
            as(qb) {
                if (typeof qb === 'function') {
                    qb = qb(new QueryBuilder());
                }
                return new Proxy(new alias.WithSubquery(qb.getSQL(), qb.getSelectedFields(), alias$1, true), new alias.SelectionProxyHandler({ alias: alias$1, sqlAliasedBehavior: 'alias', sqlBehavior: 'error' }));
            },
        };
    }
    with(...queries) {
        const self = this;
        function select(fields) {
            return new MySqlSelectBuilder({
                fields: fields ?? undefined,
                session: self.session,
                dialect: self.dialect,
                withList: queries,
            });
        }
        function selectDistinct(fields) {
            return new MySqlSelectBuilder({
                fields: fields ?? undefined,
                session: self.session,
                dialect: self.dialect,
                withList: queries,
                distinct: true,
            });
        }
        return { select, selectDistinct };
    }
    select(fields) {
        return new MySqlSelectBuilder({ fields: fields ?? undefined, session: this.session, dialect: this.dialect });
    }
    selectDistinct(fields) {
        return new MySqlSelectBuilder({
            fields: fields ?? undefined,
            session: this.session,
            dialect: this.dialect,
            distinct: true,
        });
    }
    update(table) {
        return new MySqlUpdateBuilder(table, this.session, this.dialect);
    }
    insert(table) {
        return new MySqlInsertBuilder(table, this.session, this.dialect);
    }
    delete(table) {
        return new MySqlDelete(table, this.session, this.dialect);
    }
    execute(query) {
        return this.session.execute(query.getSQL());
    }
    transaction(transaction, config) {
        return this.session.transaction(transaction, config);
    }
}
_a$1 = alias.entityKind;
MySqlDatabase[_a$1] = 'MySqlDatabase';

var _a, _b, _c;
class PreparedQuery {
}
_a = alias.entityKind;
PreparedQuery[_a] = 'MySqlPreparedQuery';
class MySqlSession {
    constructor(dialect) {
        this.dialect = dialect;
    }
    execute(query) {
        return this.prepareQuery(this.dialect.sqlToQuery(query), undefined).execute();
    }
    getSetTransactionSQL(config) {
        const parts = [];
        if (config.isolationLevel) {
            parts.push(`isolation level ${config.isolationLevel}`);
        }
        return parts.length ? alias.sql.join(['set transaction ', parts.join(' ')]) : undefined;
    }
    getStartTransactionSQL(config) {
        const parts = [];
        if (config.withConsistentSnapshot) {
            parts.push('with consistent snapshot');
        }
        if (config.accessMode) {
            parts.push(config.accessMode);
        }
        return parts.length ? alias.sql.join(['start transaction ', parts.join(' ')]) : undefined;
    }
}
_b = alias.entityKind;
MySqlSession[_b] = 'MySqlSession';
class MySqlTransaction extends MySqlDatabase {
    constructor(dialect, session, schema, nestedIndex, mode) {
        super(dialect, session, schema, mode);
        this.schema = schema;
        this.nestedIndex = nestedIndex;
    }
    rollback() {
        throw new alias.TransactionRollbackError();
    }
}
_c = alias.entityKind;
MySqlTransaction[_c] = 'MySqlTransaction';

exports.ForeignKey = ForeignKey;
exports.ForeignKeyBuilder = ForeignKeyBuilder;
exports.InlineForeignKeys = InlineForeignKeys;
exports.ManualViewBuilder = ManualViewBuilder;
exports.MySqlColumn = MySqlColumn;
exports.MySqlColumnBuilder = MySqlColumnBuilder;
exports.MySqlColumnBuilderWithAutoIncrement = MySqlColumnBuilderWithAutoIncrement;
exports.MySqlColumnWithAutoIncrement = MySqlColumnWithAutoIncrement;
exports.MySqlDatabase = MySqlDatabase;
exports.MySqlDelete = MySqlDelete;
exports.MySqlDialect = MySqlDialect;
exports.MySqlInsert = MySqlInsert;
exports.MySqlInsertBuilder = MySqlInsertBuilder;
exports.MySqlSelect = MySqlSelect;
exports.MySqlSelectBuilder = MySqlSelectBuilder;
exports.MySqlSelectQueryBuilder = MySqlSelectQueryBuilder;
exports.MySqlSession = MySqlSession;
exports.MySqlTable = MySqlTable;
exports.MySqlTransaction = MySqlTransaction;
exports.MySqlUpdate = MySqlUpdate;
exports.MySqlUpdateBuilder = MySqlUpdateBuilder;
exports.MySqlView = MySqlView;
exports.MySqlViewBase = MySqlViewBase;
exports.MySqlViewConfig = MySqlViewConfig;
exports.PreparedQuery = PreparedQuery;
exports.QueryBuilder = QueryBuilder;
exports.UniqueConstraint = UniqueConstraint;
exports.UniqueConstraintBuilder = UniqueConstraintBuilder;
exports.UniqueOnConstraintBuilder = UniqueOnConstraintBuilder;
exports.ViewBuilder = ViewBuilder;
exports.ViewBuilderCore = ViewBuilderCore;
exports.foreignKey = foreignKey;
exports.mysqlTable = mysqlTable;
exports.mysqlTableCreator = mysqlTableCreator;
exports.mysqlTableWithSchema = mysqlTableWithSchema;
exports.mysqlView = mysqlView;
exports.mysqlViewWithSchema = mysqlViewWithSchema;
exports.unique = unique;
exports.uniqueKeyName = uniqueKeyName;
